Source/WebInspectorUI/UserInterface/Views/ComputedStyleSection.js

```diff
@@ -221,6 +248,16 @@ WI.ComputedStyleSection = class ComputedStyleSection extends WI.View
             expandableView.element.classList.add("computed-property-item");
             this.element.append(expandableView.element);
         }
+
+        if (this._filterText || this._filterCleared)
+            this.dispatchEventToListeners(WI.ComputedStyleSection.Event.FilterApplied, {matches: !this.isEmpty});
+
+        this._filterCleared = false;
+    }
+
+    didLayoutSubtree()
+    {
+        this._delegate.computedStyleSectionLayoutUpdated?.(this, this.isEmpty);
```

Comment:
> Are we guaranteed to have a `this._delegate`

Yes, `WI.ComputedStyleSection` always has a delegate. Instances are created only in `WI.ComputedStyleDetailsPanel` which is the delegate passed down.

> Also, didLayoutSubtree should get called for the parent too. Do we even need a delegate call?

We still need the delegate call. I kind of prefer it.

This part suffers from the weirdness that a `WI.View` (`WI.ComputedStyleSection`) has its element manually  attached to a `WI.Object` (`WI.DetailsSectionRow`). It's not part of a continuous view tree between `WI.ComputedStyleSection` and `WI.ComputedStyleDetailsPanel`. I briefly poked it to try and make use of `didLayoutSubtree()` on the ancestor view, but it didn't work right away and I didn't investigate further. 

Given the arcane setup, I think calling `this._delegate.computedStyleSectionLayoutUpdated()` is more clear.

Reply:
> if you're not seeing didLayoutSubtree get called for the WI.ComputedStyleDetailsPanel after any WI.ComputedStyleSection does a layout then we have a bug in WI.View that absolutely should be fixed, because didLayoutSubtree should always be called after the subtree finishes layout

I don't think this is a bug.

Looking at `WI.View._visitViewTreeForLayout()`, layout is triggered only on subviews which are "dirty" (layout pending):

[Source/WebInspectorUI/UserInterface/Views/View.js?L376-378](https://sourcegraph.com/github.com/WebKit/WebKit@29b13de20953a2bdfdb1f5dfa7bf3c856d120009/-/blob/Source/WebInspectorUI/UserInterface/Views/View.js?L376-378)
```
for (let i = 0; i < views.length; ++i) {
    let view = views[i];
    if (view.layoutPending)
        view._layoutSubtree();
    else if (view._dirtyDescendantsCount)
        views.pushAll(view.subviews);
}
```

Once layout is done, the top-most subview that was "dirty" and all its subviews will indeed call `this.didLayoutSubtree()`.

[Source/WebInspectorUI/UserInterface/Views/View.js?L317-326](https://sourcegraph.com/github.com/WebKit/WebKit@29b13de20953a2bdfdb1f5dfa7bf3c856d120009/-/blob/Source/WebInspectorUI/UserInterface/Views/View.js?L317-326)

```
for (let view of this._subviews) {
    view._setLayoutReason(this._layoutReason);
    view._layoutSubtree();
}

this._layoutReason = null;

console.assert(WI.setReentrantCheck(this, "didLayoutSubtree"), "ERROR: calling `didLayoutSubtree` while already in it", this);
this.didLayoutSubtree();
```

In our case, `WI.ComputedStyleSection` is the top-most subview that does layout in response to handling the event `WI.CSSStyleDeclaration.Event.PropertiesChanged` from the computed styles object. 

The parent `WI.ComputedStyleDetailsPanel` doesn't do layout in response to this event. It does layout only when switching nodes or a `significantChange` occurs in the selected node's style.

That's why `WI.ComputedStyleDetailsPanel.didSubtreeLayout()` doesn't get called. This view isn't "dirty", but its children are.

Reply:
just because the code is currently written that way does not mean that it is incorrect

the whole idea of `didLayoutSubtree` is so that parents can do extra (ideally non-layout) work after all children have done their `layout`, meaning that the entire subtree is in a known "good" state (see the existing usage of `didLayoutSubtree` for other examples of this desired behavior)

as such, we should be calling `didLayoutSubtree` on the parent even if it's not `dirty` (i.e. it did not call `layout` on itself) so that parents can respond to changes in their children

heck one of the reasons `didLayoutSubtree` was added was to avoid this exact scenario of using an event/delegate/etc. to signal "hey im done with `layout`